### ISSUE TITLE
Is any way to highlight the selected word

### ISSUE THREAD

**Issue by rocknamx8** [jupyterlab/jupyterlab#6112]
If you are having issues with installation or configuration, you may ask for help on the [JupyterLab Gitter channel](https://gitter.im/jupyterlab/jupyterlab) or file an issue here.
Is there any jupyter lab extension support this feature. And  I have found an [extension](jcb91/jupyter_highlight_selected_word) can highlight selected word/variable, however, it only supports jupyter notebook.

**Comment by jasongrout** [comment]
No, we don't have that yet, and I don't know of any extensions that provide that feature. Perhaps that extension could be ported over to JupyterLab? (https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/highlight_selected_word/README.html, https://github.com/jcb91/jupyter_highlight_selected_word)

**Comment by Benjiini** [comment]
Are there any news on this feature?

**Comment by jasongrout** [comment]
I don't know of anyone working on this. You can highlight words across notebook editors by searching for them now (Ctrl/Cmd F)

**Comment by ramanova** [comment]
This is a basic feature which would be amazing to have. Google Collab supports it.
Plus, the search function has a bug, that takes on the focus, so unless you close the search, the page will keep jumping to the last occurrence of a searched word.

**Comment by Benjiini** [comment]
I think it would be a valuable function to debug and understand the code.
Plus it is already implemented in the Jupiter extension: nbextensions.
Im not familiar with the development of Jupiter or Jupiterlab, but is it hard to adapt this feature?

**Comment by pabloslash** [comment]
Yes please, this feature is necessary!

**Comment by sdeep27** [comment]
Another hat in the ring for this type of extensions. Would love to switch to JupyterLab fulltime, this feature would be a blessing

**Comment by jgieseler** [comment]
Very much wanted!

**Comment by jasongrout** [comment]
> Plus, the search function has a bug, that takes on the focus, so unless you close the search, the page will keep jumping to the last occurrence of a searched word.

@ramanova - this is fixed in the latest JupyterLab release

**Comment by jasongrout** [comment]
Is there anyone that wants to work on this? @jcb91, are you interested in porting https://github.com/jcb91/jupyter_highlight_selected_word to JupyterLab? We're happy to help answer questions, etc.

For everyone that wants this: the way to move forward is to work on it, or reach out to persuade someone (or even better, fund someone) to work on this. We're happy to help someone get started, answer questions, etc. We value feedback, but just posting a "me too!" comment doesn't move things forward very much.

**Comment by jcb91** [comment]
@jasongrout I'm afraid I haven't had much time recently to look at any jupyter-related stuff, but I'd certainly like to, I guess. I also haven't looked at porting anything to JupyetrLab (which of course has a significantly different extension system), so it'd probably take me a while to get up to speed on how all that works. I'll try to put some time into figuring out how jupyterlab extensions work this weekend...

**Comment by jasongrout** [comment]
Awesome. We're happy to help with any questions. And no pressure - one of the good things about open source is someone else who has more time or need can take your code as guidance and develop it themselves.

**Comment by hxtruong** [comment]
> @jasongrout I'm afraid I haven't had much time recently to look at any jupyter-related stuff, but I'd certainly like to, I guess. I also haven't looked at porting anything to JupyetrLab (which of course has a significantly different extension system), so it'd probably take me a while to get up to speed on how all that works. I'll try to put some time into figuring out how jupyterlab extensions work this weekend...
Hi, how can this amazing thing like that? I see your code is javascript. I guess there are some api of jupyter-lab to do that.

**Comment by krassowski** [comment]
Just FYI, https://github.com/krassowski/jupyterlab-lsp allows to highlights references for variable/function names under the cursor. 

It does not highlight matching words in markdown cells though (yet).

**Comment by thesofakillers** [comment]
Btw, I think this issue is a duplicate of #3104

**Comment by drorcohengithub** [comment]
Still sorely missing this :(

**Comment by josephrocca** [comment]
Since JupyterLab is using CodeMirror, implementing this should be as easy as including/importing the official `search/match-highlighter.js` addon (search for "match-highlighter" on [this page](https://codemirror.net/3/doc/manual.html)) and then adding the `highlightSelectionMatches` property when instantiating the CodeMirror instance:
```js
highlightSelectionMatches: { 
      minChars: 2,
      showToken: /\w/,
      annotateScrollbar: true,
      style: 'custom-cm-matchhighlight'
}
````
The match selection style can then be changed by adding CSS rules for the `custom-cm-matchhighlight` class. E.g.:
```css
.CodeMirror-focused .custom-cm-matchhighlight {
  background-color: #d3d3ff; /* this would need to change depending on the theme, so should use something like var(--cm-matchhighlight-color) instead */
}
```

[Here's the official CodeMirror demo](https://codemirror.net/3/demo/matchhighlighter.html) of this addon. Note that it highlights matches even if the caret is just sitting within a word, and not selecting it. This matches Google Colab behavior, and it seems reasonable so long as the default highlight color is subtle-ish (because otherwise it could be annoying). I'm not really opinionated  here though - double-clicking on a variable to select it and thus highlight other instances is easy enough.

@jasongrout @blink1073 @fcollonval Is this easy enough to do, or is a change like this entangled with a lot of other features that would require careful consideration? If it's reasonably simple to integrate the above code, which files would need to be edited, and where would `match-highlighter.js` need to be placed? Thanks!

**Comment by fcollonval** [comment]
I am not sure using the addon of CodeMirror is the right path as it is linked to the addon to custom the scrollbar that will not be expected (aka in cell highlight vs notebook highlight). This is probably the reason behind the implementation from scratch in the classical notebook extension.

I think it makes sense to add this in the document search extension as this is basically doing the same but without prompting the UI. So it seems we could take advantage of all the searching logic already provided there and adding a way to call that logic from the current CodeMirror selection.

pinging @krassowski is there some part of this problem addressed by the LSP extension?

**Comment by kangsan0420** [comment]
So, is nobody still interested in this?

**Comment by Oleg-dM** [comment]
Using for jupyter notebook for 2 years now, waiting for this highlight feature to be available in Lab ! I cannot work efficiently without having variables highlighted.. Too bad considering how cool jupyter lab looks !

**Comment by boyiisme** [comment]
wish have this feature assp

**Comment by m-beau** [comment]
up?

**Comment by seralouk** [comment]
Any news on this?

**Comment by Bortrex** [comment]
so sad! I think I'll keep using Colab connected to local runtime.

**Comment by FlorianLindinger** [comment]
I can't switch to Labs until highlighting is implemented :(

**Comment by dkobayas-cyber** [comment]
The highlighting feature seems to be available for Jupyter Lab 4.1.8 that I'm using! Thank you for those who implemented this! But I'd like to change the color of the highlighting to something more obvious. Can anyone tell me how to do that?

**Comment by DrNickBailey** [comment]
I've just upgraded to 4.2.3 and there is no highlighting of selected word that I can see or find in the options. How did you enable this @dkobayas-cyber ?

I've also come across a load of new oddities - like how line numbers are not on by default despite the option apparently set.

**Comment by segnaro** [comment]
sooo, any news on this?

**Comment by LovelyA72** [comment]
It's 2025, still no news on this feature?

**Comment by ctcjab** [comment]
I recently discovered that https://github.com/jupyter-lsp/jupyterlab-lsp gives you this feature (along with many others) and the many Jupyter users I support have been very happy since I installed it for them.
